File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = {}
        for num in nums:
            if num not in count:
                count[num] = 0
            count[num] += 1
        
        count = sorted(count.items(), key=lambda x: x[1], reverse=True)

        output = []

        for i in range(k):
            output.append(count[i][0])

        return output

        
        
        # count = {}
        # for num in nums:
        #     count[num] = count.get(num, 0) + 1

        # buckets = [[] for _ in range(len(nums) + 1)]
        # for val, freq in count.items():
        #     buckets[freq].append(val)

        # result = []
        # for freq in range(len(buckets) - 1, 0, -1):
        #     for val in buckets[freq]:
        #         result.append(val)
        #         if len(result) == k:
        #             return result

        # return result        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.ts
function topKFrequent(nums: number[], k: number): number[] {
    const count = new Map();
    for (const num of nums) count.set(num, (count.get(num) || 0) + 1);

    const buckets = Array.from({length: nums.length + 1}, () => []);
    for (const [val, freq] of count) buckets[freq].push(val);

    const result = [];
    for (let freq = buckets.length - 1; freq > 0 && result.length < k; freq--){
        for (const val of buckets[freq]){
            result.push(val)
            if (result.length === k) break;
        }
    }
    return result;
};